server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/get/TransportGetTaskAction.java

```diff
@@ -130,19 +138,33 @@ void getRunningTaskFromNode(Task thisTask, GetTaskRequest request, ActionListene
             getFinishedTaskFromIndex(thisTask, request, listener);
         } else {
             if (request.getWaitForCompletion()) {
-                // Shift to the generic thread pool and let it wait for the task to complete so we don't block any important threads.
-                threadPool.generic().execute(new AbstractRunnable() {
-                    @Override
-                    protected void doRun() {
-                        taskManager.waitForTaskCompletion(runningTask, waitForCompletionTimeout(request.getTimeout()));
-                        waitedForCompletion(thisTask, request, runningTask.taskInfo(clusterService.localNode().getId(), true), listener);
-                    }
-
-                    @Override
-                    public void onFailure(Exception e) {
-                        listener.onFailure(e);
+                final ListenableActionFuture<Void> future = new ListenableActionFuture<>();
+                RemovedTaskListener removedTaskListener = task -> {
+                    if (task.equals(runningTask)) {
+                        future.onResponse(null);
                     }
-                });
+                };
+                taskManager.registerRemovedTaskListener(removedTaskListener);
+                final ActionListener<Void> waitedForCompletionListener = ActionListener.runBefore(
+                    ActionListener.wrap(
+                        () -> waitedForCompletion(
+                            thisTask,
+                            request,
+                            runningTask.taskInfo(clusterService.localNode().getId(), true),
+                            listener
+                        )
+                    ),
+                    () -> taskManager.unregisterRemovedTaskListener(removedTaskListener)
+                );
+                future.addListener(
+                    new ThreadedActionListener<>(threadPool.executor(ThreadPool.Names.MANAGEMENT), false, waitedForCompletionListener)
```

Comment:
No need to fork here, this action immediately calls `client.get` which forks before it does anything interesting.